test/tlsconnection.js

```diff
@@ -0,0 +1,1160 @@
+    const serverHello = testHelpers.makePlaintextRecord({
+      content: await testHelpers.makeServerHelloMessage({
+        extensions: [
+          testHelpers.makeSupportedVersionsExtension(0x0304)
+        ]
+      }),
+      type: 22
+    });
+    await assert.throwsAsync(async () => {
+      await client.recv(serverHello);
+    }, TLSError, 'MISSING_EXTENSION');
+  });
+
+  it('errors if ServerHello did not select the right PSK', async () => {
+    const serverHello = testHelpers.makePlaintextRecord({
+      content: await testHelpers.makeServerHelloMessage({
+        extensions: [
+          testHelpers.makePreSharedKeyExtension(1),
+          testHelpers.makeSupportedVersionsExtension(0x0304)
+        ]
+      }),
+      type: 22
+    });
+    await assert.throwsAsync(async () => {
+      await client.recv(serverHello);
+    }, TLSError, 'ILLEGAL_PARAMETER');
+  });
+
+  it('errors if ServerHello tried to use a non-zero compression method', async () => {
+    const serverHello = testHelpers.makePlaintextRecord({
+      content: await testHelpers.makeServerHelloMessage({
+        compressionMethod: 0x01
+      }),
+      type: 22
+    });
+    await assert.throwsAsync(async () => {
+      await client.recv(serverHello);
+    }, TLSError, 'ILLEGAL_PARAMETER');
+  });
+
+  it('errors if ServerHello did not echo our sessionId', async () => {
+    const serverHello = testHelpers.makePlaintextRecord({
+      content: await testHelpers.makeServerHelloMessage({
+        sessionId: zeros(32)
+      }),
+      type: 22
+    });
+    await assert.throwsAsync(async () => {
+      await client.recv(serverHello);
+    }, TLSError, 'ILLEGAL_PARAMETER');
+  });
+
+  it('errors if the ServerHello contained unexpected extensions', async () => {
+    const serverHello = testHelpers.makePlaintextRecord({
+      content: await testHelpers.makeServerHelloMessage({
+        extensions: [
+          testHelpers.makeSupportedVersionsExtension(0x0304),
+          testHelpers.makePskKeyExchangeModesExtension([]),
+          testHelpers.makePreSharedKeyExtension(0)
+        ]
+      }),
+      type: 22
+    });
+    await assert.throwsAsync(async () => {
+      await client.recv(serverHello);
+    }, TLSError, 'UNSUPPORTED_EXTENSION');
+  });
+
+  it('errors if the ServerHello contained duplicate extensions', async () => {
+    const serverHello = testHelpers.makePlaintextRecord({
+      content: await testHelpers.makeServerHelloMessage({
+        extensions: [
+          testHelpers.makeSupportedVersionsExtension(0x0304),
+          testHelpers.makeSupportedVersionsExtension(0x0305),
+          testHelpers.makePreSharedKeyExtension(0)
+        ]
+      }),
+      type: 22
+    });
+    await assert.throwsAsync(async () => {
+      await client.recv(serverHello);
+    }, TLSError, 'DECODE_ERROR');
+  });
+
+  it('errors if ServerHello and EncryptedExtensions appear in the same record', async () => {
```

Comment:
Do you have a test for a partial record being present after the ServerHello or Finished?  I see that you have the code for checking this, but I can't find the test.

Reply:
No, I don't believe I do - perhaps that check is being exercised accidentally by some other test.  I'll add one.